File: Yoserusu/Game/Authe.cpp
#include "Authe.h"
#include "Mashiro/Mashiro.h"
#include "Mashiro/Kinect/KinectManager.h"
#include "Mashiro/Graphics/GraphicsManager.h"
#include "Mashiro/Graphics/SpriteManager.h"
#include "Mashiro/Math/Vector3.h"
#include "Util/DataBase.h"
#include "Util/DepthSingleton.h"
#include <sstream>

#include "Lua/LuaManager.h"

#include "Mashiro/Input/InputManager.h"
#include "Mashiro/Input/Keyboard.h"
#include "Mashiro/Input/Mouse.h"

using namespace std;

using namespace Mashiro;
using namespace Mashiro::Graphics;
using namespace Mashiro::Math;
using namespace Mashiro::Kinect;

Authe::Authe() : mHandCheckCount( 0 ), mHandCheck( false ), mDepth( 0 ), mCount( 0 ), mSum( 0 ),  mMoveSceneFlag( MODE_WAIT ){
	mMoveSceneFlag = MODE_WAIT;

	mBitmap[ 0 ] = Bitmap::create( "res/image/Auth.png" );
	mBitmap[ 1 ] = Bitmap::create( "res/image/Ok.png" );
	mBitmap[ 2 ] = Bitmap::create( "res/image/RedMaru.png" );
	mBitmap[ 3 ] = Bitmap::create( "res/image/BlueMaru.png" );  
	mBitmap[ 4 ] = Bitmap::create( "res/image/Title_Auth.png");

	LuaManager::instance()->loadLua( "Lua/Authe.lua" );
}

Authe::~Authe() {
}

void Authe::draw()  {
	Graphics::Manager m = Graphics::Manager::instance();
	Mashiro::Graphics::Sprite sp = Mashiro::Graphics::Sprite::instance();

	Vector2 posR = Mashiro::Kinect::Manager::instance().skeletonPos( Mashiro::Kinect::SKELETON_INDEX_HAND_RIGHT );
	Vector2 posL = Mashiro::Kinect::Manager::instance().skeletonPos( Mashiro::Kinect::SKELETON_INDEX_HAND_LEFT );

	LuaManager::instance()->runLua< int >( "Authe", "handDraw", boost::make_tuple( posR.x, posR.y, posL.x, posL.y ) );

	LuaManager::instance()->runLua< int >( "Authe", "draw", boost::make_tuple( (int)mMoveSceneFlag, check, 0, 0 ) );

}

bool Authe::handCheck()  {
	//枠ない収まってるかチェックしてます
	bool rCheck = HandRang( Mashiro::Kinect::Manager::instance().skeletonPos( Mashiro::Kinect::SKELETON_INDEX_HAND_RIGHT ), Vector2( 600, 100 ) );
	bool lCheck = HandRang( Mashiro::Kinect::Manager::instance().skeletonPos( Mashiro::Kinect::SKELETON_INDEX_HAND_LEFT ),  Vector2( 200, 100 ) );
	
	if ( rCheck && lCheck ){
		mMoveSceneFlag = MODE_NOW;
		mSum += Mashiro::Kinect::Manager::instance().depthSkeleton( Mashiro::Kinect::SKELETON_INDEX_HAND_RIGHT );
		++mCount;
		check = true;
	}else if(!rCheck && !lCheck){
		check = false;
		mCount = 0;
		mDepth = 0;
	}

	if( mCount > 20 ){
		return mHandCheck = true;
	}
	
	return mHandCheck = false;
}

void Authe::depthCheck()  {
	//深度値をうまく取ってこれるまでループさせて
	mDepth = mSum / mCount;
	if(mDepth < 15){
		mHandCheck = false;
		mSum = 0;
		mCount = 0;
		mMoveSceneFlag = MODE_NOW;
		return;
	}
	
	DepthSingleton::instance()->setDepthMin( static_cast< float >( mDepth - 3));
	DepthSingleton::instance()->setDepthMax(
		static_cast< float >(
		Mashiro::Kinect::Manager::instance().depthSkeleton(Mashiro::Kinect::SKELETON_INDEX_HEAD)) );
	
	if(DepthSingleton::instance()->getDepthMin() > DepthSingleton::instance()->getDepthMax() - DataBase::instance()->maxDepth() ){
		mHandCheck = false;
		mSum = 0;
		mCount = 0;
		mMoveSceneFlag = MODE_NOW;
		return;
	}
	if(DepthSingleton::instance()->getDepthMax()-DepthSingleton::instance()->getDepthMin() < DataBase::instance()->minDepth() ){
		mHandCheck = false;
		mSum = 0;
		mCount = 0;
		mMoveSceneFlag = MODE_NOW;
		return;
	}

	int sub = DepthSingleton::instance()->getDepthMax() - DepthSingleton::instance()->getDepthMin();
	if(abs(sub) < 15){
		mHandCheck = false;
		mSum = 0;
		mCount = 0;
		mMoveSceneFlag = MODE_NOW;
		return;
	}

	//モードをNow→END
	if(mHandCheck){
		mMoveSceneFlag = MODE_END;
	}
}

bool Authe::HandRang( const Vector2& kinectPos, const Vector2& checkPos1 ) {
	if(mHandCheckCount > 0) mHandCheckCount--;
	const float lb = checkPos1.x - 200;
	const float rb = checkPos1.x + 200;
	if( ( kinectPos.x < rb ) && ( kinectPos.x > lb ) ){
		const float tb = checkPos1.y - 200;
		const float bb = checkPos1.y + 200;
		if( ( kinectPos.y < bb ) && ( kinectPos.y > tb ) ){
			mHandCheckCount += 2 ;
			return true;
		}
	}
	return false;
}

AutheMode Authe::getMode() const {
	return mMoveSceneFlag;
}

bool Authe::handFlag() const {
	return mHandCheck;
}

File: Yoserusu/Lua/Authe.lua

require( "Lua/Enum"  );

function Authe()
    local this = {}

    this.mLeftHand = SpriteUtil();
    this.mRightHand = SpriteUtil();

    this.mAuth = SpriteUtil( "res/image/Auth.png" );
    this.mOK = SpriteUtil( "res/image/Ok.png" );
    this.mRedMaru = SpriteUtil( "res/image/RedMaru.png" );
    this.mBlueMaru = SpriteUtil( "res/image/BlueMaru.png" );  
    this.mTitle_Auth = SpriteUtil( "res/image/Title_Auth.png");
    this.mBackGraound = SpriteUtil( "res/image/bg.png" );
    this.mHandCheckCount = 0;
    
    this.draw = function( self, mMoveSceneFlag, check )
        local switch = {}

        --îwåiÇÃï`âÊ
        self.mBackGraound:draw( 0, 0 );

        --îFèÿîÕàÕâÊëú
        self.mRedMaru:draw( 600, 100 );
        self.mBlueMaru:draw( 200, 100 );

        --switchÉPÅ[ÉX 
        switch[ Mode.MODE_WAIT ] = function ()
            self.mTitle_Auth:draw( 130, 400 );
        end
        switch[ Mode.MODE_NOW ] = function ()
           self.mTitle_Auth:draw( 130, 400 );
           if ( check ) then
                self.mAuth:draw( 200, 530 );
            end
        end
        switch[ Mode.MODE_END ] = function ()
            self.mOK:draw( 130, 530 );
        end

        switch[ mMoveSceneFlag ]();

        return 0;
        
    end

    this.handDraw = function( self, rX, rY, lX, lY )
        --éËÇÃï`âÊ

        red = Vector3( 0.0, 0.0, 1.0 );

        self.mLeftHand:setColor( red );
        self.mLeftHand:drawEllipse( lX, lY, 30, 30 );

        self.mRightHand:setColor( 1.0, 0.0, 0.0 );
        self.mRightHand:drawEllipse( rX, rY, 30, 30 );

        return 0;
    end

    return this;

end